lib/utils/benchmark/src/utilsBenchmark.cpp

```diff
@@ -14,27 +14,146 @@ See the License for the specific language governing permissions and
+    }
+    state.SetComplexityN(state.range(0));
+}
+
+// Template function for benchmarking
+template <typename PassedType>
+static void BM_SubscriptorOperatorFind(benchmark::State& state) {
+    PassedType dictionary;
+    // Setup the dictionary
+    for (int i = 0; i < state.range(0); ++i) {
+        dictionary[i] = i;
+    }
+
+    for (auto _ : state) {
+		for (int i = 0; i < state.range(); i++)
+ 			benchmark::DoNotOptimize(dictionary[i]);
+    }
+    state.SetComplexityN(state.range(0));
+}
+
+// Template function for benchmarking
+template <typename PassedType>
+static void BM_Find(benchmark::State& state) {
+    PassedType dictionary;
+    // Setup the dictionary
+    for (int i = 0; i < state.range(0); ++i) {
+        dictionary[i] = i;
+    }
+
+    for (auto _ : state) {
+		for (int i = 0; i < state.range(); i++)
+ 			benchmark::DoNotOptimize(dictionary.find(i));
+    }
+    state.SetComplexityN(state.range(0));
+}
+
+// Template function for benchmarking
+template <typename PassedType>
+static void BM_Erase(benchmark::State& state) {
+    PassedType dictionary;
+    // Setup the dictionary
+    for (int i = 0; i < state.range(0); ++i) {
+        dictionary[i] = i;
+    }
+
+    for (auto _ : state) {
 		for (int i = 0; i < state.range(); i++)
-			benchmark::DoNotOptimize(a += i);
-	}
-	state.SetComplexityN(state.range(0));
+ 			benchmark::DoNotOptimize(dictionary.erase(i));
+    }
+    state.SetComplexityN(state.range(0));
 }
 
-BENCHMARK(forCycleBenchmark)
-		->RangeMultiplier(2)
-		->Range(minRange, maxRange)
-		->Complexity();
+// Register template with a specific class based on a string input
+template <typename PassedType>
+void RegisterBenchmarkSuite(const std::string& name) {
+    std::string benchmark_name = "BM_SubscriptOperatorInsert_" + name;
+    BENCHMARK_TEMPLATE(BM_SubsciptorOperatorInsert, PassedType)
+        ->Name(benchmark_name.c_str())
+        ->RangeMultiplier(2)
+        ->Range(minRange, maxRange)
+        ->Complexity();
 
-BENCHMARK_MAIN();
+	benchmark_name = "BM_Insert_" + name;
+    BENCHMARK_TEMPLATE(BM_Insert, PassedType)
+        ->Name(benchmark_name.c_str())
+        ->RangeMultiplier(2)
+        ->Range(minRange, maxRange)
+        ->Complexity();
+
+	benchmark_name = "BM_SubscriptorOperatorFind_" + name;
+    BENCHMARK_TEMPLATE(BM_SubscriptorOperatorFind, PassedType)
+        ->Name(benchmark_name.c_str())
+        ->RangeMultiplier(2)
+        ->Range(minRange, maxRange)
+        ->Complexity();
+	
+	benchmark_name = "BM_Find_" + name;
+    BENCHMARK_TEMPLATE(BM_Find, PassedType)
+        ->Name(benchmark_name.c_str())
+        ->RangeMultiplier(2)
+        ->Range(minRange, maxRange)
+        ->Complexity();
+	
+	benchmark_name = "BM_Erase_" + name;
+    BENCHMARK_TEMPLATE(BM_Erase, PassedType)
+        ->Name(benchmark_name.c_str())
+        ->RangeMultiplier(2)
+        ->Range(minRange, maxRange)
+        ->Complexity();
+}
+
+int main(int argc, char** argv) {
```

Comment:
why do you need to write a custom main?

Reply:
Because I wanted to enable the user to choose the type to run the benchmarks on. And I was not sure how to do that using BENCHMARK_MAIN